crates/example-types/src/block_types.rs

```diff
@@ -150,7 +149,7 @@ impl BlockPayload for TestBlockPayload {
         (Self::genesis(), ())
     }
 
-    fn encode(&self) -> Result<Self::Encode<'_>, Self::Error> {
+    fn encode(&self) -> Result<impl AsRef<[u8]> + Send, Self::Error> {
```

Comment:
Looks fine to me overall, but just wondering -- is there any reason we need the more general `impl AsRef<[u8]> + Send` rather than just e.g. `Arc<[u8]>`?

Reply:
Not really. Some of us have kinda started using `AsRef<[u8]>` instead of `&[u8]` just to maximize generality. I don't have a strong opinion here and could be convinced to do something else, including just `&[u8]`.

Reply:
I don't have a strong opinion either.

Is it fine if the result of `encode` gets cast to a `Vec<u8>`? Would this still avoid the collect/copy? If so, I think the `impl` is okay.

Reply:
> Is it fine if the result of `encode` gets cast to a `Vec<u8>`? Would this still avoid the collect/copy?

Uh no. The only way to return a `Vec<u8>` without copying data is to consume self.

Reply:
Then I think this should return a concrete non-Vec type, because e.g. in https://github.com/EspressoSystems/HotShot/pull/2964/files#diff-afe45f19786400fa4404b4255749941ded407b70ba5a1f26a3e12e70dfc76e08R216-R222 I believe the result gets cast to Vec<u8>